design/one-pager-pluggable-secret-stores.md

```diff
@@ -0,0 +1,267 @@
+
+Configuration of this type would require passing a typed reference to another
+configuration custom resource defined by the deployed secret store plugin.
+
+```yaml
+apiVersion: gcp.crossplane.io/v1alpha1
+kind: StoreConfig
+metadata:
+  name: vault
+spec:
+  type: External
+  defaultScope: crossplane-system
+  external:
+    endpoint: unix:///@crossplane/secret/default.sock  # The default, can be changed.
+    configRef:
+      apiVersion: secrets.crossplane.io/v1alpha1
+      kind: VaultConfig
+      name: vault-internal
+---
+apiVersion: secrets.crossplane.io/v1alpha1
+kind: VaultConfig
+metadata:
+  name: vault-internal
+spec:
+  server: http://vault.vault-system:8200
+  mountPath: secrets/
+  version: v2
+  auth:
+    method: Token
+    token:
+      source: Secret
+      secretRef:
+        namespace: crossplane-system
+        name: vault-token
+        key: token
+```
+
+There won't be any API changes at MR/XR/XRC level as they would still refer to
+a `StoreConfig` resource as before.
+
+### Implementation
+
+We will implement the `External` [Secret Store] type in Crossplane Runtime which
+will use a gRPC client to communicate with the plugin server using the service
+definition below:
+
+```protobuf
+syntax = "proto3";
+
+// This service defines the APIs for an External Secret Store plugin.
+service ExternalSecretStoreService {
+    rpc GetSecret(GetSecretRequest) returns (GetSecretResponse) {}
+    rpc ApplySecret(ApplySecretRequest) returns (ApplySecretResponse) {}
+    rpc DeleteKeys(DeleteKeysRequest) returns (DeleteKeysResponse) {}
+}
+
+message ConfigReference {
+    string api_version = 1;
+    string kind = 2;
+    string name = 3;
+}
+message Secret {
+    string scoped_name = 1;
+    map<string, string> metadata = 2;
+    map<string, bytes> data = 3;
+}
+message GetSecretRequest {
+    ConfigReference config = 1;
+    Secret secret = 2;
+    GetOptions options = 3;
+}
+message GetSecretResponse {
+    Secret secret = 1;
+}
+message ApplySecretRequest {
+    ConfigReference config = 1;
+    Secret secret = 2;
+    ApplyOptions options = 3;
+}
+message ApplySecretResponse {
+    bool changed = 1;
+}
+message DeleteKeysRequest {
+    ConfigReference config = 1;
+    Secret secret = 2;
+    DeleteOptions options = 3;
+}
+message DeleteKeysResponse {}
+message GetOptions {}
+message ApplyOptions {}
+message DeleteOptions {
+    bool keep_empty_secret = 1;
+}
+```
+
+A _Secret Store Plugin_:
+
+- Will be a binary packaged as a container image.
+- Should run a gRPC server listening at Unix domain socket.
+- May support one or more Secret Stores.
+- Should deploy CRD(s) for supported store configurations at startup.
+- Will run as a sidecar next to Crossplane and Provider containers.
+
+#### Plugin Binary vs Secret Store Implementation
+
+All `rpc` methods have `ConfigReference` as their first parameter. So,
+one Secret Store Plugin may support more than one Secret Store implementation,
```

Comment:
I like that we're deferring the decision of how many places the secret content will end up in to the plugin. I assume their CRDs would likely allow toggling that configuration.

Reply:
Their own CRD could have a config with multiple endpoint, i.e. the shape of `VaultConfig` is up to the plugin author. So, if there is a use case for multiple places, then the plugin could have that list in its own config CR instance.